File: patterns/creational/prototype.py
"""
Prototype Pattern - Python

Creates objects by cloning existing instances (prototypes) rather than
creating new instances from scratch.

Use cases:
- When object creation is expensive
- When you want to avoid subclassing
- When you need to create objects at runtime
"""

import copy
from abc import ABC, abstractmethod


# Prototype interface
class Shape(ABC):
    @abstractmethod
    def clone(self):
        pass
    
    @abstractmethod
    def draw(self):
        pass


# Concrete prototypes
class Circle(Shape):
    def __init__(self, radius, color):
        self.radius = radius
        self.color = color
    
    def clone(self):
        return copy.deepcopy(self)
    
    def draw(self):
        return f"Drawing a {self.color} circle with radius {self.radius}"


class Rectangle(Shape):
    def __init__(self, width, height, color):
        self.width = width
        self.height = height
        self.color = color
    
    def clone(self):
        return copy.deepcopy(self)
    
    def draw(self):
        return f"Drawing a {self.color} rectangle {self.width}x{self.height}"


# Prototype registry
class ShapeRegistry:
    def __init__(self):
        self.prototypes = {}
    
    def register(self, key, prototype):
        self.prototypes[key] = prototype
    
    def create(self, key):
        prototype = self.prototypes.get(key)
        if not prototype:
            raise ValueError(f"Prototype {key} not found")
        return prototype.clone()


# Example usage
if __name__ == "__main__":
    print("=== Prototype Pattern ===\n")
    
    # Create original shapes
    original_circle = Circle(10, 'red')
    original_rectangle = Rectangle(20, 30, 'blue')
    
    print("Original shapes:")
    print(original_circle.draw())
    print(original_rectangle.draw())
    
    print()
    
    # Clone shapes
    cloned_circle = original_circle.clone()
    cloned_rectangle = original_rectangle.clone()
    
    print("Cloned shapes:")
    print(cloned_circle.draw())
    print(cloned_rectangle.draw())
    
    print()
    print("Are they different objects?", original_circle is not cloned_circle)  # True
    
    # Using registry
    registry = ShapeRegistry()
    registry.register('red-circle', original_circle)
    registry.register('blue-rectangle', original_rectangle)
    
    new_circle = registry.create('red-circle')
    new_circle.radius = 15  # Modify cloned object
    print("Modified clone:", new_circle.draw())
    print("Original unchanged:", original_circle.draw())

File: patterns/creational/prototype.js
/**
 * Prototype Pattern - JavaScript
 * 
 * Creates objects by cloning existing instances (prototypes) rather than
 * creating new instances from scratch.
 * 
 * Use cases:
 * - When object creation is expensive
 * - When you want to avoid subclassing
 * - When you need to create objects at runtime
 */

// Prototype interface
class Shape {
    clone() {
        throw new Error('Method must be implemented');
    }
    
    draw() {
        throw new Error('Method must be implemented');
    }
}

// Concrete prototypes
class Circle extends Shape {
    constructor(radius, color) {
        super();
        this.radius = radius;
        this.color = color;
    }
    
    clone() {
        return new Circle(this.radius, this.color);
    }
    
    draw() {
        return `Drawing a ${this.color} circle with radius ${this.radius}`;
    }
}

class Rectangle extends Shape {
    constructor(width, height, color) {
        super();
        this.width = width;
        this.height = height;
        this.color = color;
    }
    
    clone() {
        return new Rectangle(this.width, this.height, this.color);
    }
    
    draw() {
        return `Drawing a ${this.color} rectangle ${this.width}x${this.height}`;
    }
}

// Prototype registry
class ShapeRegistry {
    constructor() {
        this.prototypes = {};
    }
    
    register(key, prototype) {
        this.prototypes[key] = prototype;
    }
    
    create(key) {
        const prototype = this.prototypes[key];
        if (!prototype) {
            throw new Error(`Prototype ${key} not found`);
        }
        return prototype.clone();
    }
}

// Example usage
console.log('=== Prototype Pattern ===\n');

// Create original shapes
const originalCircle = new Circle(10, 'red');
const originalRectangle = new Rectangle(20, 30, 'blue');

console.log('Original shapes:');
console.log(originalCircle.draw());
console.log(originalRectangle.draw());

console.log();

// Clone shapes
const clonedCircle = originalCircle.clone();
const clonedRectangle = originalRectangle.clone();

console.log('Cloned shapes:');
console.log(clonedCircle.draw());
console.log(clonedRectangle.draw());

console.log();
console.log('Are they different objects?', originalCircle !== clonedCircle); // true

// Using registry
const registry = new ShapeRegistry();
registry.register('red-circle', originalCircle);
registry.register('blue-rectangle', originalRectangle);

const newCircle = registry.create('red-circle');
newCircle.radius = 15; // Modify cloned object
console.log('Modified clone:', newCircle.draw());
console.log('Original unchanged:', originalCircle.draw());

module.exports = { Shape, Circle, Rectangle, ShapeRegistry };

